### DISCUSSION TITLE
Motion warping

### DISCUSSION THREAD

**Discussion by ywmaa** [godotengine/godot-proposals#4930] [3D]
### Describe the project you are working on

3D Advanced Movement template

### Describe the problem or limitation you are having in your project

godot doesn't have motion warping

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Motion warping is a technique to move the character to the desired location so the animations like executions, door interaction, or any kind of interaction looks good.
Example :
https://m.youtube.com/watch?v=71HIG52A6QI

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I will probably use tweens to move the character in the needed time, also add like a notify track in the animations to handle when should the motion warping happen.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

it can be worked around, I was initially going to implement it in my project :
https://github.com/ywmaa/Advanced-Movement-System-Godot/issues/17

### Is there a reason why this should be core and not an add-on in the asset library?

I thought that it could be used in all types of games, I am doing this proposal to see if most people would like to have it in directly Godot Engine, if so I may work on it.

**Comment by Calinou** [comment]
Note: Moved to discussion, as there is no concrete implementation proposed in OP.

**Reply by ywmaa** [reply]
looks right thing to do.

**Reply by ywmaa** [reply]
@Calinou 
since it now has an implementation, should it go back to issues ?

**Reply by Calinou** [reply]
"An implementation" for a proposal means that you should have a concrete idea of how to implement this in C++, or at least describe an API that would be used by people to use motion warping in their projects. Unfortunately, having an add-on alone is not sufficient here.

cc @TokageItLab in case they have opinions on motion warping :slightly_smiling_face:

**Reply by ywmaa** [reply]
> "An implementation" for a proposal means that you should have a concrete idea of how to implement this in C++, or at least describe an API that would be used by people to use motion warping in their projects. Unfortunately, having an add-on alone is not sufficient here.
> 
> cc @TokageItLab in case they have opinions on motion warping 🙂

Ok, but shouldn't the idea be discussed first before doing any implementation ?  "we will have a motion warping node in the Godot source code" 

Or motion warping is already needed to be in Godot core ? 

Excuse my misunderstanding, it is almost the first time for me to try to contribute, and from the docs I understand that issues should be discussed before an implementation 
Does it get discussed here or in an issue ?

**Comment by ywmaa** [comment]
here is the example code, I didn't test it yet, but it should work
https://github.com/ywmaa/Advanced-Movement-System-Godot/blob/master/addons/PoseWarping/MotionWarping.gd

**Comment by ywmaa** [comment]
Tested it, 
![Screenshot from 2022-07-23 22-00-08](https://user-images.githubusercontent.com/68971808/180621461-2ee402e3-1d4f-4a7a-aa34-b7c389d84dde.png)


https://user-images.githubusercontent.com/68971808/180621498-32a5c206-925e-43c9-9121-95082628f9f0.mp4


for me it is a great way to combine the power of tweens, and AnimationPlayer, I may work on getting it directly in Godot Source Code 
it just needs a sync_rotation and I think it is perfect


here are examples what people use it for :
https://www.youtube.com/watch?v=71HIG52A6QI
https://www.youtube.com/watch?v=_zz-b75jM5Y
https://www.youtube.com/watch?v=ddd4kR2MTcA

**Comment by TokageItLab** [comment]
Is the "motion warping" feature particular to the Unreal Engine? A particular feature in a particular software is not meant to be generic. The proposal should explain "what it does", "what use cases it has", and "what procedures could be used to implement it".

From what I see in the video, is it the feature to interpolate animation or root motion to a specific target? IMO, it could be implemented as an add-on using the value track or method track, without implementing it in core.

**Reply by ywmaa** [reply]
Yeah it is introduced by unreal  engine, 

The purpose of this tech is to sync player position with the animation, you know in docs AnimationPlayer Is useful for the animations that you know already its values, and tweens are useful for values that you don't know it, yet.

Here in Godot motion warping stores the target position from the code .

And in the Animation player you know when we should arrive to the target, so you tell the motion warping to execute the "tween" in this case using the AnimationPlayer keyframe.

It is now working as an add-on, and actually it is like 40 lines of code.

I though that It can be used in all types of games, so I thought in not only sync player position, but allow the developer to sync any property he wants (yeah this is already done in a tween), and execute the tween using AnimationPlayer 

So basically this plugin stores the values and executes it.
